test/extensions/io_socket/user_space_io_socket/user_space_file_event_impl_test.cc

```diff
@@ -0,0 +1,387 @@
+  dispatcher_->run(Event::Dispatcher::RunType::NonBlock);
+}
+
+TEST_F(UserSpaceFileEventImplTest, RescheduleAfterTriggered) {
+  user_file_event_ = std::make_unique<UserSpaceFileEventImpl>(
+      *dispatcher_, [this](uint32_t arg) { ready_cb_.called(arg); }, event_rw, io_source_);
+  {
+    SCOPED_TRACE("1st schedule");
+    user_file_event_->activate(event_rw);
+    EXPECT_CALL(ready_cb_, called(event_rw));
+    dispatcher_->run(Event::Dispatcher::RunType::NonBlock);
+  }
+
+  {
+    SCOPED_TRACE("2nd schedule");
+    user_file_event_->activate(event_rw);
+    EXPECT_CALL(ready_cb_, called(event_rw));
+    dispatcher_->run(Event::Dispatcher::RunType::NonBlock);
+  }
+  {
+    SCOPED_TRACE("merge events");
+    user_file_event_->activate(Event::FileReadyType::Read);
+    user_file_event_->activate(Event::FileReadyType::Write);
+    EXPECT_CALL(ready_cb_, called(event_rw));
+    dispatcher_->run(Event::Dispatcher::RunType::NonBlock);
+  }
+}
+
+TEST_F(UserSpaceFileEventImplTest, RescheduleIsDeduplicated) {
+  user_file_event_ = std::make_unique<UserSpaceFileEventImpl>(
+      *dispatcher_, [this](uint32_t arg) { ready_cb_.called(arg); }, event_rw, io_source_);
+  {
+    SCOPED_TRACE("1st schedule");
+    user_file_event_->activate(event_rw);
+
+    user_file_event_->activate(event_rw);
+    EXPECT_CALL(ready_cb_, called(event_rw));
+    dispatcher_->run(Event::Dispatcher::RunType::NonBlock);
+  }
+
+  {
+    SCOPED_TRACE("further dispatcher drive");
+    EXPECT_CALL(ready_cb_, called(_)).Times(0);
+    dispatcher_->run(Event::Dispatcher::RunType::NonBlock);
+  }
+}
+
+TEST_F(UserSpaceFileEventImplTest, DefaultReturnAllEnabledReadAndWriteEvents) {
+  for (const auto current_event : {Event::FileReadyType::Read, Event::FileReadyType::Write,
+                                   Event::FileReadyType::Read | Event::FileReadyType::Write}) {
+    SCOPED_TRACE(absl::StrCat("current event:", current_event));
+    EXPECT_CALL(io_source_, isReadable())
+        .WillOnce(Return((current_event & Event::FileReadyType::Read) != 0))
+        .RetiresOnSaturation();
+    EXPECT_CALL(io_source_, isPeerWritable())
+        .WillOnce(Return((current_event & Event::FileReadyType::Write) != 0))
+        .RetiresOnSaturation();
+    auto user_file_event = std::make_unique<UserSpaceFileEventImpl>(
+        *dispatcher_, [this](uint32_t arg) { ready_cb_.called(arg); }, event_rw, io_source_);
+    EXPECT_CALL(ready_cb_, called(current_event));
+    dispatcher_->run(Event::Dispatcher::RunType::NonBlock);
+  }
+}
+
+TEST_F(UserSpaceFileEventImplTest, ActivateWillSchedule) {
+  // IO is neither readable nor writable.
+  user_file_event_ = std::make_unique<UserSpaceFileEventImpl>(
+      *dispatcher_, [this](uint32_t arg) { ready_cb_.called(arg); }, event_rw, io_source_);
+  {
+    EXPECT_CALL(ready_cb_, called(_)).Times(0);
+    dispatcher_->run(Event::Dispatcher::RunType::NonBlock);
+  }
+  {
+    user_file_event_->activate(Event::FileReadyType::Read);
```

Comment:
Would be good to verify that activate works even in cases where the mask passed in to setEnabled does not contain the event being activated.

Reply:
That's a critical case. It may be hidden in some test cases. Let me create a new named test case